File: integration-tests-java/src/test/java/io/micronaut/jimmer/java/it/event/TestChangeEventObserves.java
package io.micronaut.jimmer.java.it.event;

import io.micronaut.jimmer.java.it.entity.Book;
import io.micronaut.jimmer.java.it.entity.BookProps;
import io.micronaut.jimmer.java.it.entity.BookStoreProps;
import io.micronaut.runtime.event.annotation.EventListener;
import jakarta.inject.Singleton;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import org.babyfish.jimmer.sql.event.AssociationEvent;
import org.babyfish.jimmer.sql.event.EntityEvent;

@Singleton
public class TestChangeEventObserves {

    private final List<EntityEvent<?>> entityEventStorage = new CopyOnWriteArrayList<>();

    private final List<AssociationEvent> associationEventStorageOne = new CopyOnWriteArrayList<>();

    private final List<AssociationEvent> associationEventStorageTwo = new CopyOnWriteArrayList<>();

    @EventListener
    public void entityChangeEvent(EntityEvent<?> entityEvent) {
        if (entityEvent.getImmutableType().getJavaClass() == Book.class) {
            entityEventStorage.add(entityEvent);
        }
    }

    @EventListener
    public void associationChangeEvent(AssociationEvent associationEvent) {
        if (associationEvent.isChanged(BookProps.STORE)) {
            associationEventStorageOne.add(associationEvent);
        } else if (associationEvent.isChanged(BookStoreProps.BOOKS)) {
            associationEventStorageTwo.add(associationEvent);
        }
    }

    public List<EntityEvent<?>> getEntityEventStorage() {
        return entityEventStorage;
    }

    public List<AssociationEvent> getAssociationEventStorageOne() {
        return associationEventStorageOne;
    }

    public List<AssociationEvent> getAssociationEventStorageTwo() {
        return associationEventStorageTwo;
    }
}

File: integration-tests-kotlin/src/test/kotlin/io/micronaut/jimmer/kotlin/it/event/TestChangeEventObserves.kt
package io.micronaut.jimmer.kotlin.it.event

import io.micronaut.jimmer.kotlin.it.entity.Book
import io.micronaut.jimmer.kotlin.it.entity.BookProps
import io.micronaut.jimmer.kotlin.it.entity.BookStoreProps
import io.micronaut.runtime.event.annotation.EventListener
import jakarta.inject.Singleton
import org.babyfish.jimmer.sql.event.AssociationEvent
import org.babyfish.jimmer.sql.event.EntityEvent
import java.util.concurrent.CopyOnWriteArrayList

@Singleton
class TestChangeEventObserves {
    private val entityEventStorage: MutableList<EntityEvent<*>> = CopyOnWriteArrayList()

    private val associationEventStorageOne: MutableList<AssociationEvent> = CopyOnWriteArrayList()

    private val associationEventStorageTwo: MutableList<AssociationEvent> = CopyOnWriteArrayList()

    @EventListener
    fun entityChangeEvent(entityEvent: EntityEvent<*>) {
        if (entityEvent.getImmutableType().getJavaClass() == Book::class.java) {
            entityEventStorage.add(entityEvent)
        }
    }

    @EventListener
    fun associationChangeEvent(associationEvent: AssociationEvent) {
        when {
            associationEvent.isChanged(BookProps.STORE) -> associationEventStorageOne.add(associationEvent)
            associationEvent.isChanged(BookStoreProps.BOOKS) -> associationEventStorageTwo.add(associationEvent)
        }
    }

    fun getEntityEventStorage(): MutableList<EntityEvent<*>> = entityEventStorage

    fun getAssociationEventStorageOne(): MutableList<AssociationEvent> = associationEventStorageOne

    fun getAssociationEventStorageTwo(): MutableList<AssociationEvent> = associationEventStorageTwo
}
